MaxText/sft/hooks.py

```diff
@@ -0,0 +1,177 @@
+#  Copyright 2025 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+"""Training and data loading hooks for SFT"""
+
+from flax import nnx
+from collections import defaultdict
+import datetime
+import jax
+import jax.numpy as jnp
+
+from MaxText.metric_logger import MetricLogger
+from MaxText.utils import gcs_utils
+from MaxText import exceptions
+from MaxText import maxtext_utils
+from MaxText import max_logging
+from MaxText import max_utils
+from MaxText.data_loader import DataLoader
+from MaxText.input_pipeline.input_pipeline_interface import create_data_iterator
+from tunix.sft import peft_trainer
+from tunix.sft.hooks import DataHooks, TrainingHooks
+from typing_extensions import override
+from MaxText.utils.goodput_utils import (
+    GoodputEvent,
+    record_goodput,
+)
+
+
+class SFTTrainingHooks(TrainingHooks):
+  """Training hooks for SFT."""
+
+  def __init__(self, config, mesh, learning_rate_schedule, goodput_recorder):
+    self.config = config
+    self.mesh = mesh
+    self.metric_logger = MetricLogger(self.config, learning_rate_schedule)
+    self.goodput_recorder = goodput_recorder
+    self.metadata = {
+        "first_train_step": None,
+    }
+    self.train_metadata = defaultdict(float)
+    self.eval_metadata = defaultdict(float)
+
+  @override
+  def on_train_start(self, train_ctx: peft_trainer.PeftTrainer):
+    """Called at the beginning of training."""
+    state = nnx.state(train_ctx.model)
+    params = state.filter(nnx.Param)
+
+    if not self.config.using_pipeline_parallelism:
+      maxtext_utils.assert_params_sufficiently_sharded(params, self.mesh, self.config.sharding_tolerance)
+
+    self.metric_logger.write_setup_info_to_tensorboard(params)
+
+    if self.config.dump_hlo:
+      jax.block_until_ready(state)  # Ensure compilation has finished
+      gcs_utils.upload_dump(
+          self.config.dump_hlo_local_dir,
+          self.config.dump_hlo_gcs_dir,
+          module_name=self.config.dump_hlo_module_name,
+          delete_local_after=self.config.dump_hlo_delete_local_after,
+          all_host_upload=self.config.dump_hlo_upload_all,
+      )
+
+    self.metadata["first_train_step"] = train_ctx.train_steps
+
+  @override
+  def on_train_end(self, train_ctx: peft_trainer.PeftTrainer):  # pylint: disable=unused-argument
```

Comment:
Use your judgement on this but I have a vague feeling that we ought to assert that `.*_start` is called before `.*_end` to catch bugs from the Tunix side which would be quite nasty if hit. I wouldn't go as far as creating new state to check but I think here there's enough state already that you could assert on. But your call, just an itch I'm feeling